### PR TITLE
setting up a basic e2e test to verify component CR get created

### PR DESCRIPTION
**Pull request by corinnekrych**
fixed odc-132
This PR adds:
- a simple e2e test for deploying `devopsconsole-operator` and creating a component CR component
- add Makefile support for testing locally in dev mode (test are not run in container) for easy writing and debugging of the 2e2 tests

### CODE REVIEW

File: make/test.mk [hunk 1]
```diff
@@ -120,12 +120,12 @@ test-e2e: build build-image e2e-setup
...
 e2e-setup:  e2e-cleanup
-	oc new-project devconsole-e2e-test || true
+	@-oc new-project devconsole-e2e-test
 
 .PHONY: e2e-cleanup
 e2e-cleanup:
-	oc login -u system:admin
-	oc delete project devconsole-e2e-test || true
+	@-oc login -u system:admin
+	@-oc delete project devconsole-e2e-test
```
Comment (tinakurian) [inline make/test.mk]: should we maybe cleanup the other resources we created (crds, etc)?

Comment (ldimaggi) [reply make/test.mk]: We will want the test to always have a "clean" starting point, so that it can be run repeatedly.

**Comment by tinakurian** [comment]
I have fixed the intermittent issues related to the test failing if run consecutively. This test still needs a bit of work - not all the resources are cleaned up once the test has executed.

File: test/e2e/component_operator_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"context"
+	"fmt"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	"github.com/operator-framework/operator-sdk/pkg/test/e2eutil"
+	"github.com/redhat-developer/devopsconsole-operator/pkg/apis"
+	componentsv1alpha1 "github.com/redhat-developer/devopsconsole-operator/pkg/apis/devopsconsole/v1alpha1"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"os"
```
[the remark this answers is not in the document]
Comment (tinakurian) [reply test/e2e/component_operator_test.go]: @baijum please verify this is fixed properly!

[the remark this answers is not in the document]
Comment (baijum) [reply test/e2e/component_operator_test.go]: Looks good.

**Comment by corinnekrych** [comment]
@tinakurian indeed Tina those are created by component controller and should be cleared.

File: test/e2e/component_operator_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	t.Log("Initialized cluster resources")
+
+	namespace, err := ctx.GetNamespace()
+	require.NoError(t, err, "failed to get namespace where operator needs to run")
+
+	// get global framework variables
+	f := framework.Global
+	t.Log(fmt.Sprintf("namespace: %s", namespace))
+	// wait for component-operator to be ready
+	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "devopsconsole-operator", 1, retryInterval, timeout)
```
Comment (Avni-Sharma) [inline test/e2e/component_operator_test.go]: The namespace here should be "operators", I believe

Comment (corinnekrych) [reply test/e2e/component_operator_test.go]: @Avni-Sharma Indeed.
That should be the namespace in https://github.com/redhat-developer/devopsconsole-operator/pull/34/files#diff-153541cef4d3666ae354851a6f6ef7c2R8 
@tinakurian we should still have 2 namespaces:
- one where we deploy the operator (could be "operators" as suggested by Avni, on the one i've used devopsconsole)
- one where a use wants to deploy its component CR, as one requirement is to be able to make sure we can watch CR from _any_ namespace, I'd rather have 2 different one.

I think with the changed you did we're atm use only one namespace (which should work too) but to reflect reality better have 2 different one.

Comment (corinnekrych) [reply test/e2e/component_operator_test.go]: @tinakurian after further thoughts, I think we better keep that first PR simple, pls clean the resources and then we merge that one. 
We can then add more tests.

File: make/test.mk [hunk 4]
```diff
@@ -345,3 +345,21 @@ CLEAN_TARGETS += clean-coverage-unit
...
+#-------------------------------------------------------------------------------
+
+.PHONY: build-image-local
+build-image-local: e2e-setup
+	eval $$(minishift docker-env) && operator-sdk build $(shell minishift openshift registry)/devconsole-e2e-test/devopsconsole-operator
+
+.PHONY: e2e-local
+e2e-local: build-image-local
+	@-oc project devconsole-e2e-test
+	@-oc apply -f deploy/crds/devopsconsole_v1alpha1_component_crd.yaml
```
[the remark this answers is not in the document]
Comment (tinakurian) [reply make/test.mk]: @baijum fixed here: 7292817b029570c12b651c255982dbd7e851c25d

File: make/test.mk (hunk 1 above)
Comment (tinakurian) [reply make/test.mk]: fixed here: 36723998ea7fb932dd460587d4f242edf8029148

File: make/test.mk [hunk 5]
```diff
@@ -345,3 +355,39 @@ CLEAN_TARGETS += clean-coverage-unit
...
+e2e-local: build-image-local
+	@-oc login -u system:admin
+	@-oc project $(TEST_NAMESPACE)
+	@-oc create -f $(CUR_DIR)/deploy/crds/devopsconsole_v1alpha1_component_crd.yaml
+	@-oc create -f $(CUR_DIR)/deploy/service_account.yaml --namespace $(TEST_NAMESPACE)
+	@-oc create -f $(CUR_DIR)/deploy/role.yaml --namespace $(TEST_NAMESPACE)
+ifeq ($(UNAME_S),Darwin)
+	@sed -i "" 's|REPLACE_NAMESPACE|$(TEST_NAMESPACE)|g' $(CUR_DIR)/deploy/test/role_binding_test.yaml
+else
+	@sed -i 's|REPLACE_NAMESPACE|$(TEST_NAMESPACE)|g' $(CUR_DIR)/deploy/test/role_binding_test.yaml
```
Comment (corinnekrych) [inline make/test.mk]: @tinakurian this syntax is wont fit in macOS env there fore my contribution here
01d3344 to make it work on mac.

**Comment by corinnekrych** [comment]
Approved. 
If you're ok @tinakurian we can merge that one. It's a good first step.

Next steps cc @ldimaggi :
- integrate e2e  test to CI rework `make/test.mk` `test-e2e` target, clean up and update test/global-manifest.yaml 
- modify the set up in makefile to create 2 namespaces one to deploy operator controllers (maybe linked to olm deployment) and another namespace to deploy CR, we need to make sure we test RBAC properly.
- add more e2e tests (edge cases) for now we have only one flow covered: create a component CR and check it created associated resources (buildconfig, imagestream, build)
- ... and more I'm sure ;)

File: make/test.mk (hunk 5 above)
Comment (ldimaggi) [reply make/test.mk]: Corrine read my mind about being able to run on Mac or Linux - approved. 

(Thx to Corrine for additional test ideas too!)

File: make/test.mk [hunk 6]
```diff
@@ -87,7 +87,10 @@ ALL_PKGS_EXCLUDE_PATTERN = 'vendor\|app\|tool\/cli\|design\|client\|test'
 # This pattern excludes some folders from the go code analysis
 GOANALYSIS_PKGS_EXCLUDE_PATTERN="vendor|app|client|tool/cli"
 GOANALYSIS_DIRS=$(shell go list -f {{.Dir}} ./... | grep -v -E $(GOANALYSIS_PKGS_EXCLUDE_PATTERN))
+RANDOM := $(shell bash -c 'echo $$RANDOM')
 
+export TEST_NAMESPACE='devconsole-e2e-test-$(RANDOM)'
```
Comment (sbose78) [inline make/test.mk]: Good :)

**Comment by sbose78** [comment]
> not all the resources are cleaned up

Do we know why? @corinnekrych @tinakurian 
Garbage collection following the `ownerReferences` should be taking care of it.

**Comment by tinakurian** [comment]
> > not all the resources are cleaned up
> 
> Do we know why? @corinnekrych @tinakurian
> Garbage collection following the `ownerReferences` should be taking care of it.

We weren't deleting the namespace after the test had run - it's taken care of now in func TearDownSuite()

File: make/test.mk [hunk 7]
```diff
@@ -87,7 +87,10 @@ ALL_PKGS_EXCLUDE_PATTERN = 'vendor\|app\|tool\/cli\|design\|client\|test'
 # This pattern excludes some folders from the go code analysis
 GOANALYSIS_PKGS_EXCLUDE_PATTERN="vendor|app|client|tool/cli"
 GOANALYSIS_DIRS=$(shell go list -f {{.Dir}} ./... | grep -v -E $(GOANALYSIS_PKGS_EXCLUDE_PATTERN))
+RANDOM := $(shell bash -c 'echo $$RANDOM')
```
[the remark this answers is not in the document]
Comment (sbose78) [reply make/test.mk]: Yup, dashes are allowed.

[the remark this answers is not in the document]
Comment (tinakurian) [reply make/test.mk]: fixed here: 8fc26b5f73b3e2f46dffe23421eacbadb27304ae

[the remark this answers is not in the document]
Comment (sbose78) [reply make/test.mk]: .. though I'm slightly concerned about long namespace names. So, please do double-check.

[the remark this answers is not in the document]
Comment (baijum) [reply make/test.mk]: From:
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names

`By convention, the names of Kubernetes resources should be up to maximum length of 253 characters and consist of lower case alphanumeric characters, -, and ., but certain resources have more specific restrictions.`

So, this is fine.

File: make/test.mk (hunk 4 above)
[the remark this answers is not in the document]
Comment (kwk) [reply make/test.mk]: @baijum we don't need to abandon relative paths as we're ALWAYS building inside the source directory. That is common for Go projects and I don't see when this will ever change.
